File: benchmarks/pairs/vibe-py/pythonssti-main-read-root/fixed.py
# Provenance: TheWation/PythonSSTI read_root (fixed).
# repo: TheWation/PythonSSTI
# commit: 76339c27d5a870c0a81bad6aee5f9bd516592145
# parent: 76339c27d5a870c0a81bad6aee5f9bd516592145
# commit_url: https://github.com/TheWation/PythonSSTI/blob/76339c27d5a870c0a81bad6aee5f9bd516592145/main.py#L14
# cve: 
# license: MIT
# function: read_root
# relpath: main.py
# provenance: human
# mechanism: source_reaches_sink

async def read_root(username=None):

    username = username or "Guest"
    Jinja2 = Environment()

    # Vulnerable Implementation
    output = Jinja2.from_string("Welcome " + username + "!").render()

    # Safe Implementation
    # output = Jinja2.from_string("Welcome {{ user_name }}!").render(user_name=username)

    return HTMLResponse(content=output)

File: benchmarks/pairs/vfc-js/dns-sync-0-1-0/fixed.js
// Provenance: skoranga/node-dns-sync  (fixed).
// repo: skoranga/node-dns-sync
// commit: d9abaae384b198db1095735ad9c1c73d7b890a0d
// parent: b748fe77a87749d95876028eda8e759109be4c73
// commit_url: https://github.com/skoranga/node-dns-sync/commit/d9abaae384b198db1095735ad9c1c73d7b890a0d
// cve: CVE-2014-9682
// license: MIT
// function: resolve
// relpath: lib/dns-sync.js
// provenance: human
// mechanism: source_reaches_sink
// upstream_start: 19

module.exports = {
    resolve: function resolve(hostname) {
        var output,
            nodeBinary = process.execPath;

        if (!isValidHostName(hostname)) {
            console.error('Invalid hostname:', hostname);
            return null;
        }

        var scriptPath = path.join(__dirname, "../scripts/dns-lookup-script"),
            response,
            cmd = util.format('"%s" "%s" %s', nodeBinary, scriptPath, hostname);

        response = shell.exec(cmd, {silent: true});
        if (response && response.code === 0) {
            output = response.output;
            if (output && net.isIP(output)) {
                return output;
            }
        }
        debug('hostname', "fail to resolve hostname " + hostname);
        return null;
    }
};
